### ISSUE TITLE
[CustomTextEditor extensions] Random error "file has changed in the meantime" due to sync vs async mismatch

### ISSUE THREAD

**Issue by Ark-kun** [microsoft/vscode#149016]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.66.2
- OS Version: Windows

TL/DR:
When implementing CustomTextEditor extension, text edits randomly fail since `vscode.workspace.applyEdit` requires synchronicity, but `webviewPanel.webview.onDidReceiveMessage` cannot provide synchronicity.

Steps to Reproduce:

```typescript
export class MyEditorProvider implements vscode.CustomTextEditorProvider {
  constructor(private readonly context: vscode.ExtensionContext) {}

  public async resolveCustomTextEditor(
    document: vscode.TextDocument,
    webviewPanel: vscode.WebviewPanel,
    _token: vscode.CancellationToken
  ): Promise<void> {
    // ...
    webviewPanel.webview.onDidReceiveMessage(async (message) => {
      const edit = new vscode.WorkspaceEdit();
      edit.replace(
        document.uri,
        new vscode.Range(0, 0, document.lineCount, 0),
        message
      );
      await vscode.workspace.applyEdit(edit);
    };
  }
```

Then load into webview some HTML that frequently posts messages to VSCode.

You'll start getting errors:
```
workbench.desktop.main.js:610   ERR Error: file:///c%3A/.../testData/pipeline2.component.yaml has changed in the meantime
    at i._validateBeforePrepare (workbench.desktop.main.js:893:19404)
    at i.apply (workbench.desktop.main.js:893:20155)
    at u._performTextEdits (workbench.desktop.main.js:2666:96950)
    at u.perform (workbench.desktop.main.js:2666:96219)
    at apply (workbench.desktop.main.js:2666:98428)
    at D.$tryApplyWorkspaceEdit (workbench.desktop.main.js:2673:4103)
    at d._doInvokeHandler (workbench.desktop.main.js:1705:13100)
    at d._invokeHandler (workbench.desktop.main.js:1705:12784)
    at d._receiveRequest (workbench.desktop.main.js:1705:11446)
    at d._receiveOneMessage (workbench.desktop.main.js:1705:10124)
    at workbench.desktop.main.js:1705:8233
    at S.invoke (workbench.desktop.main.js:67:145)
    at s.fire (workbench.desktop.main.js:67:1856)
    at MessagePort.F.onmessage (workbench.desktop.main.js:1961:52056)
workbench.desktop.main.js:610  WARN IGNORING workspace edit Error: file:///c%3A/.../testData/pipeline2.component.yaml has changed in the meantime
    at i._validateBeforePrepare (workbench.desktop.main.js:893:19404)
    at i.apply (workbench.desktop.main.js:893:20155)
    at u._performTextEdits (workbench.desktop.main.js:2666:96950)
    at u.perform (workbench.desktop.main.js:2666:96219)
    at apply (workbench.desktop.main.js:2666:98428)
    at D.$tryApplyWorkspaceEdit (workbench.desktop.main.js:2673:4103)
    at d._doInvokeHandler (workbench.desktop.main.js:1705:13100)
    at d._invokeHandler (workbench.desktop.main.js:1705:12784)
    at d._receiveRequest (workbench.desktop.main.js:1705:11446)
    at d._receiveOneMessage (workbench.desktop.main.js:1705:10124)
    at workbench.desktop.main.js:1705:8233
    at S.invoke (workbench.desktop.main.js:67:145)
    at s.fire (workbench.desktop.main.js:67:1856)
    at MessagePort.F.onmessage (workbench.desktop.main.js:1961:52056)
```

The error seems to be caused by `vscode.workspace.applyEdit` being called before the previous call has completed. So, `vscode.workspace.applyEdit` requires synchronicity. But unfortunately `webviewPanel.webview.onDidReceiveMessage` cannot provide synchronicity. It does not `await` the `async` handlers. This makes me think that the problem is not fixable for the extention developer.

**Comment by mjbvz** [comment]
Without a better example, it sounds like all the apis are working as expected (`onDidReceiveMessage` is not directly tied to editing)

If you are seeing an issue  with your extension, try sequencing the edits itself by awaiting previous edits before trying to commit new ones

**Comment by Ark-kun** [comment]
@mjbvz
>onDidReceiveMessage is not directly tied to editing

In a sense, there is a direct tie. CustomTextEditor is based on Webview. The **only** way the webview can send the document modification is via `onDidReceiveMessage`. There is no other way. So, in CustomTextEditor, the edits are always applied inside `onDidReceiveMessage`.

>try sequencing the edits yourself by awaiting previous edits
In the code I've included, I am awaiting `vscode.workspace.applyEdit`. But it does not help.

Would do you think about making `vscode.workspace.applyEdit` async-compatible? The system owns the editing service, so that's where the necessary sequencing should probably take place.

Think about the issue again from a different angle: The issue is that calling `applyEdit` has a random chance of failing.
There is no way to know whether calling `applyEdit` is safe to call.
There is no `await vscode.workspace.waitForLastEditCompletion()`.
Using `try/catch` does not help, since, again, we cannot wait for the service to become ready.

I think the proper solution is to make `applyEdit` apply edits without throwing errors (by sequencing edits or any other means).
